src/mscorlib/src/System/Collections/Generic/DefaultObjectPool.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Collections.Generic
+{
+    internal class DefaultObjectPool<T> : ObjectPool<T> where T : class, new()
+    {
+        /// <summary>The default maximum number of oebjects that are available for rent.</summary>
+        private const int DefaultMaxPooled = 128;
+
+        private readonly T[] _objects;
+
+        private SpinLock _lock; // do not make this readonly; it's a mutable struct
+        private int _index;
+
+        /// <summary>
+        /// Creates the pool with DefaultMaxPooled objects.
+        /// </summary>
+        public DefaultObjectPool()
+            : this(DefaultMaxPooled)
+        {
+        }
+
+        /// <summary>
+        /// Creates the pool with maxPooled objects.
+        /// </summary>
+        public DefaultObjectPool(int maxPooled)
+        {
+            if (maxPooled <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxPooled));
+            }
+
+            _lock = new SpinLock(Debugger.IsAttached); // only enable thread tracking if debugger is attached; it adds non-trivial overheads to Enter/Exit
+            _objects = new T[maxPooled];
+        }
+
+        /// <summary>Takes an object from the pool. If the pool is empty, returns a new object.</summary>
+        public override T Rent()
+        {
+            T[] objects = _objects;
+            T obj = null;
+
+            // While holding the lock, grab whatever is at the next available index and
+            // update the index.  We do as little work as possible while holding the spin
+            // lock to minimize contention with other threads.  The try/finally is
+            // necessary to properly handle thread aborts on platforms which have them.
```

Comment:
This code isn't thread abort safe if you care about that. `index` may be incremented without setting the array element to null. Then in `Return` `index` can be decremented without setting the array element to the returned object. Next call to `Rent` will return an object that's still in use.